src/hooks/utility/prepare.rs

```diff
@@ -0,0 +1,208 @@
+        let mut types_oid: *mut pg_sys::Oid = if nargs > 0 {
+            let oid_ptr = pg_sys::palloc((nargs as usize) * std::mem::size_of::<pg_sys::Oid>())
+                as *mut pg_sys::Oid;
+            let type_elements = (*arg_types).elements;
+            for i in 0..(*arg_types).length {
+                let type_name =
+                    (*type_elements.offset(i as isize)).ptr_value as *const pg_sys::TypeName;
+                *oid_ptr.offset(i as isize) = pg_sys::typenameTypeId(pstate, type_name)
+            }
+            oid_ptr
+        } else {
+            null_mut()
+        };
+
+        #[cfg(any(feature = "pg15", feature = "pg16", feature = "pg17"))]
+        {
+            pg_sys::parse_analyze_varparams(
+                &mut raw_stmt,
+                (*pstate).p_sourcetext,
+                &mut types_oid,
+                &mut nargs,
+                null_mut(),
+            )
+        }
+        #[cfg(any(feature = "pg13", feature = "pg14"))]
+        {
+            pg_sys::parse_analyze_varparams(
+                &mut raw_stmt,
+                (*pstate).p_sourcetext,
+                &mut types_oid,
+                &mut nargs,
+            )
+        }
+    };
+
+    let query_relations = get_query_relations(unsafe { (*query).rtable });
+    if unsafe { (*query).commandType } != pg_sys::CmdType::CMD_SELECT
+        || !is_duckdb_query(&query_relations)
+    {
+        return Ok(true);
+    }
+
+    // set search path according to postgres
+    set_search_path_by_pg()?;
+
+    let query_sql: &CStr = unsafe { CStr::from_ptr((*pstate).p_sourcetext) };
+    if let Err(e) = connection::execute(query_sql.to_str()?, []) {
+        fallback_warning!(e.to_string());
+        return Ok(true);
+    }
+
+    // It's always necessary to execute the previous hook to store a prepared statement in PostgreSQL
+    Ok(true)
+}
+
+pub fn execute_query<T: pgbox::WhoAllocated>(
+    _psate: *mut pg_sys::ParseState,
+    stmt: *mut pg_sys::ExecuteStmt,
+    query_desc: PgBox<pg_sys::QueryDesc, T>,
+) -> Result<bool> {
+    unsafe {
+        let prepared_stmt = pg_sys::FetchPreparedStatement((*stmt).name, true);
+        let plan_source = (*prepared_stmt).plansource;
+
+        if plan_source.is_null() || !(*plan_source).fixed_result {
+            return Ok(true);
+        }
+
+        // we need to make Duckdb replan the `PREPARE` statement when search path changed.
+        #[cfg(not(feature = "pg17"))]
+        let need_replan = !pg_sys::OverrideSearchPathMatchesCurrent((*plan_source).search_path);
+
+        #[cfg(feature = "pg17")]
+        let need_replan = !pg_sys::SearchPathMatchesCurrentEnvironment((*plan_source).search_path);
+
+        #[cfg(feature = "pg13")]
+        let cached_plan = pg_sys::GetCachedPlan(plan_source, null_mut(), false, null_mut());
+
+        #[cfg(not(feature = "pg13"))]
+        let cached_plan = pg_sys::GetCachedPlan(plan_source, null_mut(), null_mut(), null_mut());
```

Comment:
Why do you need to call GetCachedPlan twice here? Could you please add a comment to explain?

Reply:
GetCachedPlan interface changed in PG14. so here is a cfg. Add a simple comment